akka-persistence-testkit/src/test/scala/akka/persistence/testkit/scaladsl/UnpersistentDurableStateSpec.scala

```diff
@@ -0,0 +1,257 @@
+      probe.expectPersisted(State(1, Map.empty, Int.MaxValue), tag = "count")
+      assert(!testkit.hasEffects(), "should have no actor effects")
+    }
+
+    "allow a state to be injected" in {
+      import BehaviorUnderTest._
+
+      val behavior = BehaviorUnderTest("test-1", TestInbox[Done]().ref)
+
+      val notify3 = TestInbox[Done]()
+      val initialState = State(1, Map(3 -> notify3.ref), 3)
+
+      val (unpersistent, probe) =
+        UnpersistentBehavior.fromDurableState[Command, State](behavior, Some(initialState))
+
+      val testkit = BehaviorTestKit(unpersistent)
+      val logs = testkit.logEntries()
+      logs.size shouldBe 1
+      logs.head.level shouldBe Level.DEBUG
+      logs.head.message shouldBe s"Recovered state for id [test-1] is [$initialState]"
+      assert(!probe.hasEffects, "should be no persistence effect")
+      assert(!notify3.hasMessages, "no messages should be sent to notify3")
+
+      val replyTo = TestInbox[Done]()
+      testkit.run(AddWhenAtLeast(2, 2, replyTo.ref))
+      assert(!replyTo.hasMessages, "no messages should be sent now")
+      assert(!notify3.hasMessages, "no messages should be sent to notify3")
+      assert(!probe.hasEffects, "should be no persistence effect")
+      assert(!testkit.hasEffects(), "should be no testkit effects")
+
+      testkit.run(Add(3, TestInbox[Done]().ref))
+      replyTo.expectMessage(Done)
+      notify3.expectMessage(Done)
+      assert(!testkit.hasEffects(), "should be no testkit effects")
+      probe.drain() should contain theSameElementsInOrderAs Seq(
+        PersistenceEffect(State(4, Map.empty, Int.MaxValue), 1, Set("count")),
+        PersistenceEffect(State(6, Map.empty, Int.MaxValue), 2, Set("count")))
+    }
+
+    "stash and unstash properly" in {
+      import BehaviorUnderTest._
+
+      val behavior = BehaviorUnderTest("test-1", TestInbox[Done]().ref)
+
+      val (unpersistent, probe) = UnpersistentBehavior.fromDurableState[Command, State](behavior)
+
+      val replyTo1 = TestInbox[Done]()
+      val add = Add(1, TestInbox[Done]().ref)
+      val testkit = BehaviorTestKit(unpersistent)
+
+      // stashes
+      testkit.run(AddWhenAtLeast(1, 1, replyTo1.ref))
+      assert(!probe.hasEffects, "should be no persistence effect")
+      assert(!replyTo1.hasMessages, "count is not yet 1")
+
+      // unstashes
+      testkit.run(add)
+      replyTo1.expectMessage(Done)
+      probe.drain() shouldNot be(empty)
+
+      // unstash but nothing in the stash
+      testkit.run(add)
+      assert(!replyTo1.hasMessages, "should not send again")
+      probe.drain() shouldNot be(empty)
+    }
+
+    "retrieve revision number" in {
+      import BehaviorUnderTest._
+
+      val behavior = BehaviorUnderTest("test-1", TestInbox[Done]().ref)
+
+      val (unpersistent, probe) = UnpersistentBehavior.fromDurableState[Command, State](behavior)
+
+      val replyTo = TestInbox[Long]()
+      val testkit = BehaviorTestKit(unpersistent)
```

Comment:
Is the unpersistent behavior ever going to be used for anything else than wrapping it in a behavior testkit and if so should we do that out of the box for less noise?

Reply:
The BehaviorTestKit is almost certain to be the main use-case (and that accounts for at least some of the speedup that's observed).  That said, the BTK's more fundamental limitations around testing interactions between different behaviors might motivate using this in a more async style where you still assume that persistence "just works".

For that, the ActorTestKit is possible, but the persistence probe API probably wouldn't work well (there'd be a lot of `eventually`s, I'd expect).  Since events/snapshots/states are kind of intrinsically stream-like, maybe the persistence probes could be exposed in that style as streams (though backpressure might really complicate things?).

So maybe have plumbing (returning the behavior and maybe even exposing the callbacks) and porcelain which returns a testkit and probe(s)?

Reply:
For now, just porcelain, but plumbing can be exposed pending demand.